aw-transform/src/union.rs

```diff
@@ -0,0 +1,178 @@
+
+            // If event1 starts before event2, add that event
+            if intersect_timestamp > event1.timestamp {
+                let prepended_event = Event {
+                    id: None,
+                    timestamp: event1.timestamp,
+                    duration: intersect_timestamp - event1.timestamp,
+                    data: event1.data.clone(),
+                };
+                events.push(prepended_event);
+            }
+
+            // Add intersecting event
+            let mut intersect_data = event1.data.clone();
+            merge_map(&mut intersect_data, &event2.data);
+            let intersecting_event = Event {
+                id: None,
+                timestamp: intersect_timestamp,
+                duration: intersect_duration,
+                data: intersect_data,
+            };
+            events.push(intersecting_event);
+
+            // Update event1 to end at end of common event
+            event1.timestamp = intersect_endtime;
+            event1.duration = event1_endtime - intersect_endtime;
+            if event1.duration.num_milliseconds() <= 0 {
+                continue 'event1;
+            }
+        }
+        events.push(event1);
+    }
+
+    events
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use chrono::DateTime;
+    use chrono::Duration;
+    use serde_json::json;
+    use std::str::FromStr;
+
+    #[test]
+    fn test_merge_data() {
+        /* test merge same */
+        let mut d1 = json_map! {"test": json!(1)};
+        let d2 = d1.clone();
+        merge_map(&mut d1, &d2);
+        assert_eq!(d1, d2);
+
+        /* test merge different keys */
+        let mut d1 = json_map! {"test1": json!(1)};
+        let d2 = json_map! {"test2": json!(2)};
+        merge_map(&mut d1, &d2);
+        assert_eq!(d1, json_map! {"test1": json!(1), "test2": json!(2)});
+
+        /* test merge intersecting objects */
+        let mut d1 = json_map! {"test": json_map!{"a": json!(1)}};
+        let d2 = json_map! {"test": json_map!{"b": json!(2)}};
+        merge_map(&mut d1, &d2);
+        assert_eq!(
+            d1,
+            json_map! {"test": json_map!{"a": json!(1), "b": json!(2)}}
+        );
+
+        /* test non-object conflict, prefer map1 value */
+        // TODO: This does not work yet!
+        //       It should be a pretty rare use-case anyway
+        /*
+        let mut d1 = json_map!{"test": json!(1)};
+        let d1_orig = d1.clone();
+        let d2 = json_map!{"test": json!(2)};
+        merge_map(&mut d1, &d2);
+        assert_eq!(d1, d1_orig);
+        */
+    }
+
+    #[test]
+    fn test_union_events_split() {
+        // Test intersection, before and after
+        let e1 = Event {
+            id: None,
+            timestamp: DateTime::from_str("2000-01-01T00:00:00Z").unwrap(),
+            duration: Duration::seconds(3),
+            data: json_map! {"test": json!(1)},
+        };
+        let mut e2 = e1.clone();
```

Comment:
I'd like to have more realistic data here, for example:

For e1: `{app: "firefox", title: "google - firefox"}`
For e2: `{title: "google", url: "google.com"}`

Event series that are at least two events long, like the one I recently commented about, would also be nice (would also help with ensuring behavior stays consistent when we eventually remove the nested loops).